Ensemble weighted by inverse relative skill

kathsherratt opened the pull request:
This PR adds code to ensemble by inverse relative skill, building on the `relative_skill` metric in the `evaluation/` csv files.

- Added ensemble method [`create_ensemble_relative_skill()`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/e79a32c44e767ead6e38f918c66e350ef99e1e4c/code/ensemble/methods/create-ensemble-relative-skill.R)
  - Finds an evaluation csv from an input date
  - Takes inverse of relative skill to use as weights
  - Weights input forecasts' values and sums to create ensemble forecast
  - Returns `ensemble` dataframe and optionally returns `weights` dataframe
  
- Updated ensemble infrastructure code to support running a relative skill ensemble
  - [`run_ensemble()`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/e79a32c44e767ead6e38f918c66e350ef99e1e4c/code/ensemble/utils/run-ensemble.R) supports method "relative_skill" with new args 
    - `evaluation_date` for finding evaluation csv (defaults to `forecast_date`)
    - `continuous_weeks` for selecting models only with a continuous history of `>=x`  weeks (defaults to 4+ weeks)
    - `return_criteria` returns `weights` dataframe, as well as criteria for ensemble model selection
  - [`format_ensemble()`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/e79a32c44e767ead6e38f918c66e350ef99e1e4c/code/ensemble/utils/format-ensemble.R) supports the (slightly different) output from `create_ensemble_relative_skill()` for formatting into hub standard
  - [`run_multiple_ensembles()`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/e79a32c44e767ead6e38f918c66e350ef99e1e4c/code/ensemble/utils/run-multiple-ensembles.R) supports the new args
  - `verbose` option added throughout (maybe handy for interactive use - displays forecast and evaluation dates, number of models used, errors)

- Created ensembles using weekly evaluations from 2021-04-05 onwards
  - [`create-all-methods-ensembles.R`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/e79a32c44e767ead6e38f918c66e350ef99e1e4c/code/ensemble/utils/create-all-methods-ensembles.R) includes code to run a single model for all past dates, implemented for method "relative_skill"
  - This produced relative skill ensembles over all past dates with weekly evaluation:
    - 8 ensembles in csv hub format saved in `ensembles/data-processed`
    - 8 weights for respective ensembles saved in new folder `ensembles/weights`

sbfnk commented on code/ensemble/methods/create-ensemble-relative-skill.R:
```diff
@@ -0,0 +1,77 @@
+
+  # Weights for each model, horizon, location, target
+  skill <- skill %>%
+    group_by(target_variable, location, horizon) %>%
```
can we make grouping by horizon optional? I don't think it's a bad idea as such, but having varying weights by horizon means that one might get pathological behaviour (e.g., uncertainty decreasing with increasing horizon). If not grouping by horizon one could average the skill over all horizons first by model, location and target variable and then apply the weighting by location and target variable.

sbfnk commented on code/ensemble/utils/create-all-methods-ensembles.R:
```diff
@@ -15,21 +15,59 @@ all_dates <- vroom(here("code", "ensemble", "EuroCOVIDhub",
+single_method <- "relative_skill"
+
+# Run
+ensembles <- run_multiple_ensembles(forecast_dates = all_dates,
```
Not sure what the intention here is - doesn't this repeat what was already done above when running all methods?

kathsherratt replied:
Ah - yes, it does repeat the above (except for saving weights), I was just using for testing/debugging. Removed.

sbfnk reviewed (changes requested):
Looking great!

kathsherratt commented:
Relative skill is now mean-averaged across all horizons, grouped by model, location, and target_variable, before taking the inverse and creating/applying the weights.

- Added arg `by_horizon` to [`create_ensemble_relative_skill()`](https://github.com/epiforecasts/covid19-forecast-hub-europe/blob/ensemble-relative-skill/code/ensemble/methods/create-ensemble-relative-skill.R)
  - `FALSE` (default) takes mean relative skill over all horizons for each model, location, target variable, then weights and applies to values.
  - `TRUE` takes relative skill by horizon, model, location, target_variable, then weights and applies to values by horizon.
- Using `by_horizon = FALSE`, re-created all the past relative skill ensembles (previously created with weights by horizon). 
  - This overwrites contents of `ensembles/data-processed/EuroCOVIDhub-relative_skill` (8 csvs) and `ensembles/weights/EuroCOVIDhub-relative_skill` (8 csvs)

Also removed duplicate test code from `create-all-methods-ensembles.R`
